llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -177,6 +182,81 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineSetExtract(
   return true;
 }
 
+// Returns a map with InsertIndices and registers holding the insert values.
+std::optional<std::map<unsigned, Register>>
+AIE2PreLegalizerCombinerImpl::getVectorInsertIndices(
+    MachineInstr *CurMI, unsigned SclSrcBits, MachineRegisterInfo &MRI) const {
+  std::map<unsigned, Register> RegMap;
+  auto Is8BitVInsert = [](const MachineInstr *MI) {
+    return isa<GIntrinsic>(MI) && cast<GIntrinsic>(*MI).getIntrinsicID() ==
+                                      Intrinsic::aie2_vinsert8_I512;
+  };
+  auto IsSet = [](const MachineInstr *MI) {
+    return isa<GIntrinsic>(MI) && (cast<GIntrinsic>(*MI).getIntrinsicID() ==
+                                       Intrinsic::aie2_set_I512_I128 ||
+                                   cast<GIntrinsic>(*MI).getIntrinsicID() ==
+                                       Intrinsic::aie2_set_I512_I256);
+  };
+
+  while (Is8BitVInsert(CurMI)) {
+    // In this case of G_INTRINSIC operand 1 is target intrinsic
+    const Register SrcReg = CurMI->getOperand(2).getReg();
+    const Register IdxReg = CurMI->getOperand(3).getReg();
+    const Register SclSrcReg = CurMI->getOperand(4).getReg();
+
+    // Collecting registers and their indices
+    auto Cst = getIConstantVRegValWithLookThrough(IdxReg, MRI);
+    if (!Cst ||
+        !RegMap.try_emplace(Cst->Value.getZExtValue(), SclSrcReg).second)
+      break;
+    CurMI = getDefIgnoringCopies(SrcReg, MRI);
+
+    // Combining Set and Extract to fetch next VInsert
+    if (IsSet(CurMI) && tryToCombineSetExtract(*CurMI))
+      CurMI = getDefIgnoringCopies(SrcReg, MRI);
+  }
+
+  // For 128/256-bit vectors, not all lanes are explicitly defined. If the
+  // source MI is identified as a Set intrinsic that sets the required lanes,
+  // the transformation can proceed safely.
+  if (!IsSet(CurMI))
+    return std::nullopt;
+  unsigned DstRegBits =
+      MRI.getType(CurMI->getOperand(2).getReg()).getSizeInBits();
+  // Check for the right amount of lanes matching the size of input vector of
+  // Set instrinsic.
+  if (DstRegBits != RegMap.size() * SclSrcBits)
+    return std::nullopt;
+  return RegMap;
+}
+
+/// Look for VINSERT sequence that can be rewritten as G_BUILD_VECTOR_TRUNC
+bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorInserts(
+    MachineInstr &MI, unsigned SclSrcBits) const {
+  std::map<unsigned, Register> RegMap;
+  MachineInstr *CurMI = &MI;
+
+  auto InsertIndices = getVectorInsertIndices(CurMI, SclSrcBits, MRI);
+  if (!InsertIndices)
+    return false;
+  unsigned DstRegLen = InsertIndices->size();
+
+  MachineIRBuilder MIRBuilder(MI);
+  SmallVector<Register, 16> Regs;
+  // Collect registers in order for G_BUILD_VECTOR_TRUNC
+  for (unsigned I = 0; I < DstRegLen; I++) {
+    auto It = InsertIndices->find(I);
+    if (It == InsertIndices->end())
+      return false;
+    Regs.push_back(It->second);
+  }
+  Register DstRegTrunc = MRI.createGenericVirtualRegister(
+      LLT::fixed_vector(DstRegLen, SclSrcBits));
```

Comment:
Could we re-use the type from `MI.getOperand(0)` instead of trying to reconstruct it with `LLT::fixed_vector(DstRegLen, SclSrcBits)`?

Reply:
Why is it like this actually, do our VINSERT.S8 intrinsics take a `<16 x s32>` type?

Reply:
yes, they take `<16 x s32>` type